src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -628,6 +630,50 @@ public override string GetName(int ordinal)
 
         #endregion
 
+        #region Generic value getters
+
+        internal delegate T ReadDelegate<T>(NpgsqlReadBuffer buffer, int columnLen, FieldDescription fieldDescription);
+
+        internal delegate ValueTask<T> ReadAsyncDelegate<T>(NpgsqlReadBuffer buffer, int columnLen, bool async, FieldDescription fieldDescription);
+
+        internal static class NullableHandler<T>
+        {
+            public static readonly ReadDelegate<T> Read;
+            public static readonly ReadAsyncDelegate<T> ReadAsync;
+            public static bool Exists = NullableHandler.Construct(out Read, out ReadAsync);
```

Comment:
Unfortunately, not. The JIT doesn't elide such things, but there is an issue (dotnet/coreclr#2591) in the CoreCLR repo.

Reply:
The non-generic `NullableHandler` class is a factory which holds everything it needs and exposes only the `Construct<T>` method, nothing more. If I move the construction part to the static constructor of `NullableHandler<T>` which is a holder, then I must change access modifiers of all factory members to `internal` or `public` to make them accessible from the generic class. Also it would mix purposes of these classes. So I did what I did. It's a common practice to split logic (see `Nullable` and `Nullable<T>` for example).

Reply:
Added lost `readonly`. It would improve performance because the JIT treats readonly fields as constants.

Reply:
My proposal was simpler I think... All I was thinking was to move the invocation of `Construct<T>` to a static constructor on the same class, not to change the relationship `NullableHandler` and `NullableHandler<T>`. It's just a minor readability suggestion (when I first looked at your code, I actually misread the assignment to `Exists` as a property getter...).

So instead of:

```c#
internal static class NullableHandler<T>
{
    public static readonly ReadDelegate<T> Read;
    public static readonly ReadAsyncDelegate<T> ReadAsync;
    public static readonly bool Exists = NullableHandler.Construct(out Read, out ReadAsync);
}
````

it would be:

```c#
internal static class NullableHandler<T>
{
    static NullableHandler() {
        Exists = NullableHandler.Construct(out Read, out ReadAsync);
    }
    public static readonly ReadDelegate<T> Read;
    public static readonly ReadAsyncDelegate<T> ReadAsync;
    public static readonly bool Exists;
}
````

(or something similar).

If you don't like this for any reason we can skip it.

This is my last comment, after this I can merge, let me know.

Reply:
Ah so (: It's not a problem. Will change.

Reply:
Done. Let's merge.